fairseq/data/data_utils.py

```diff
@@ -24,9 +24,25 @@ def infer_language_pair(path):
     return src, dst
 
 
-def collate_tokens(values, pad_idx, eos_idx=None, left_pad=False, move_eos_to_beginning=False):
+def get_pad_size(values, input_shapes):
+    if input_shapes is None:
+        return max(v.size(0) for v in values)
+    for batch_size, padlen in input_shapes:
+        if len(values) == batch_size:
+            return padlen
+    else:
+        raise IndexError(
+            'Encountered values with invalid length {}, input shapes were {}'
+            .format(len(values), input_shapes)
+        )
```

Comment:
This indenting is weird to me: why is the `for` loop unindented but in between the `if` and `else`? I'd probably put the `for` loop inside the `else` block unless there's some style point here I'm missing

Reply:
for/else is a python thing. else is executed if you don't early eexit out of the for loop. here, len(values) _must_ be inside input_shapes and we must early return, otherwise we raise error. I can get rid of the else, it would have the same effect, wdyt?

Reply:
cool, TIL. Looks good as-is in that case